### PR TITLE
[GLM Image] Add e2e pipeline in nightly and benchmark CI

### LINKED ISSUE #5571: [GLM Image] Add e2e pipeline in nightly and benchmark CI

### PR DESCRIPTION
**Pull request by meenakshiramanathan1**
### Ticket

- Fixes #5571 

### Problem description

To add GLM Image Text-to-image model to nightly + benchmark CI. 

### What's changed

- Adds a per-step PCC gate to the GLM-Image nightly e2e pipeline test (tests/torch/models/glm_image/test_pipeline.py), following the playground_v2_5 / sdxl_lightning e2e conventions (tt-xla#5480). GLM-Image is a diffusion text-to-image model (not next-scale autoregressive): a DiT transformer (GlmImageTransformer2DModel, 30 blocks) denoises the latent over a FlowMatchEuler schedule with classifier-free guidance.

- Nightly (tests/torch/models/glm_image/test_pipeline.py): drives the shared GlmImagePipeline from tt_forge_models (the same pipeline the imagegen benchmark uses) end-to-end. The DiT runs on TT in bf16, tensor-parallel sharded on the ("batch","model") mesh. The AR vision-language encoder (prior-token generation), the T5 glyph text encoder, the FlowMatchEuler scheduler, and the BSQ VAE decode all stay on CPU. Native target 1024×1024; run at 30 inference steps. Markers: nightly, model_test, llmbox, tensor_parallel, large, record_test_properties.

- Per-step PCC gate (nightly): the DiT is the only component on TT, so it's the only component gated. After every per-step, per-CFG-branch TT forward, the same (CPU-cast, upcast-to-fp32) inputs are fed to a lazily-loaded fp32 CPU twin of the transformer and logits/output are compared inline. _pcc computes PCC via TorchComparisonEvaluator and the check fails fast the moment any forward drops below PCC_THRESHOLD = 0.99. The pipeline itself keeps using the TT outputs (real deployment behavior); the twin is purely observational.

- Submodule left untouched. GLM's pipeline lives in the shared third_party/tt_forge_models submodule and is reused by the benchmark, so the test monkeypatches pipeline.transformer.forward after setup() rather than editing the submodule. The twin is loaded lazily on the first forward; other callers (benchmark, examples) are unchanged and the submodule carries no tt-xla test dependency.

- Here only the DiT runs on TT — the T5 encoder, vision-language encoder, and VAE only touch TT in their isolated component tests (test_text_encoder.py, test_transformer.py, test_vae_decoder.py, test_vision_language_encoder.py), so those are retained for coverage.

- Sequential classifier-free guidance (pre-existing in the pipeline): cond and uncond are run as two batch-1 forwards per step and combined on the output; this is why the 30-step run performs 60 gated forwards.

### Validation: 
All 30 steps (both CFG branches — 60 forwards) passed the 0.99 gate; the pipeline then decoded a coherent 1024×1024 image and the dimension asserts passed.

| Test | Result | Time |
| --- | --- | --- |
| `test_pipeline.py::test_infinity_pipeline` (nightly e2e + per-scale PCC) | ✅ passed | 1h08m07s (4087.67s) |

Per-step DiT PCC vs. fp32 CPU twin (min = 0.999727, step 5 uncond; max = 0.999896):

| Scale | cond | uncond |
| --- | --- | --- |
| 1–10 | 0.999844–0.999889| 0.999727–0.999891 |
| 11-20 | 0.999862–0.999892 | 0.999813–0.999896 |
| 21-30 | 0.999814–0.999891 | 0.999807–0.999876| 

Final image generated from pipeline:
glm_image_pipeline_output_jul14

### Checklist

- [x] Validate nightly e2e + benchmark on qb2 blackhole

- [x] Perf benchmark CI run (glm_image, qb2 blackhole): https://github.com/tenstorrent/tt-xla/actions/runs/29722400701

### CODE REVIEW

**Review by vvukomanTT — APPROVED** [review]

**Review by nsumrakTT — APPROVED** [review]

File: tests/benchmark/benchmarks/glm_image_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+        def _to_tt(x, dtype=None):
+            if not on_tt:
+                return x
+            if dtype is not None:
+                x = x.to(dtype)
+            return x.to(xm.xla_device())
+
+        def _to_cpu(x):
+            return x.to("cpu") if on_tt else x
```
Comment (vkovacevicTT) [inline tests/benchmark/benchmarks/glm_image_pipeline.py:119]: I would drop these functions and cast directly.

File: tests/benchmark/test_imagegen.py [hunk 2]
```diff
@@ -457,3 +411,51 @@ def test_janus_pro_7b(output_file, request):
...
     )
+
+
+def test_glm_image(output_file, request):
+    """GLM-Image diffusion text-to-image benchmark (DiT tensor-parallel).
+
+    Unlike the single-chip SD models above, GLM-Image's DiT transformer runs
+    tensor-parallel across a multi-chip mesh (the AR vision-language encoder, T5
+    glyph encoder, FlowMatchEuler scheduler and VAE decode stay on CPU). The
+    matrix pins this entry to an llmbox (n300-llmbox) runner so the mesh is available.
```
Comment (vkovacevicTT) [inline tests/benchmark/test_imagegen.py:518]: NIT: We don't use n300-llmbox anymore. Change to qb2-blackhole here.

File: tests/benchmark/benchmarks/glm_image_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    GlmImageConfig,
+)
+from third_party.tt_forge_models.glm_image.pytorch.src.pipeline import (
+    GlmImagePipeline as _BaseGlmImagePipeline,
+)
+
+__all__ = ["GlmImageConfig", "GlmImagePipeline", "PROMPT", "SEED", "HEIGHT", "WIDTH"]
+
+
+class GlmImagePipeline(_BaseGlmImagePipeline):
```
Comment (vkovacevicTT) [inline tests/benchmark/benchmarks/glm_image_pipeline.py:50]: `mesh_shape` isn't set anywhere. It silently passes None.

https://github.com/tenstorrent/tt-xla/blob/7963edc84bc90b00d13c9653a1203842776e4e30/tests/benchmark/benchmarks/imagegen_benchmark.py#L205

In perf report of job you linked in Checklist, it is `null` for qb2-blackhole.

**Review by vkovacevicTT — APPROVED** [review]

File: tests/benchmark/test_imagegen.py (hunk 2 above)
Comment (meenakshiramanathan1) [reply tests/benchmark/test_imagegen.py:518]: Modified

File: tests/torch/models/lumina_image/test_transformer.py [hunk 4]
```diff
@@ -14,15 +14,14 @@
 
 from third_party.tt_forge_models.lumina_image.pytorch import ModelLoader, ModelVariant
 
+# @pytest.mark.skip
```
Comment (sgligorijevicTT) [inline tests/torch/models/lumina_image/test_transformer.py]: This can be deleted rather than commented out

File: tests/torch/models/lumina_image/test_transformer.py [hunk 5]
```diff
@@ -14,15 +14,14 @@
...
 
-@pytest.mark.skip
-def test_transformer():
-    _run(sharded=False)
 
-
-@pytest.mark.xfail(
-    reason="loc(concatenate.459): error: failed to legalize unresolved materialization - https://github.com/tenstorrent/tt-xla/issues/4901"
-)
+# @pytest.mark.xfail(
```
Comment (sgligorijevicTT) [inline tests/torch/models/lumina_image/test_transformer.py]: This can be deleted rather than commented out

**Review by sgligorijevicTT — APPROVED** [review]
